From 654a2ce19927e752c8a7858288ac9049a580b43c Mon Sep 17 00:00:00 2001
From: Simon Desfarges <simonx.desfarges@intel.com>
Date: Wed, 17 Sep 2014 12:24:51 +0200
Subject: [PATCH 367/429] spi: set the real bus frequency in chip structure

This patch allows the user to get the real clock frequency of the bus.

Signed-off-by: Simon Desfarges <simonx.desfarges@intel.com>
---
 drivers/spi/intel_mid_ssp_spi.c | 14 ++++++++++++--
 1 file changed, 12 insertions(+), 2 deletions(-)

diff --git a/drivers/spi/intel_mid_ssp_spi.c b/drivers/spi/intel_mid_ssp_spi.c
index 24e7267c8fb8..70aa59ec17ea 100644
--- a/drivers/spi/intel_mid_ssp_spi.c
+++ b/drivers/spi/intel_mid_ssp_spi.c
@@ -920,6 +920,15 @@ static unsigned int ssp_get_clk_div(struct ssp_drv_context *sspc, int speed)
 		return clamp(100000000 / speed - 1, 3, 4095);
 }
 
+
+static int ssp_get_speed(struct ssp_drv_context *sspc, int clk_div)
+{
+	if (sspc->quirks & QUIRKS_PLATFORM_MRFL)
+		return 25000000 / (clk_div + 1);
+	else
+		return 100000000 / (clk_div + 1);
+}
+
 /**
  * transfer() - Start a SPI transfer
  * @spi:	Pointer to the spi_device struct
@@ -1283,9 +1292,10 @@ static int setup(struct spi_device *spi)
 	}
 
 	if ((sspc->quirks & QUIRKS_SPI_SLAVE_CLOCK_MODE) == 0) {
-		chip->speed_hz = spi->max_speed_hz;
-		clk_div = ssp_get_clk_div(sspc, chip->speed_hz);
+		clk_div = ssp_get_clk_div(sspc, spi->max_speed_hz);
 		chip->cr0 |= (clk_div & 0xFFF) << 8;
+		spi->max_speed_hz = ssp_get_speed(sspc, clk_div);
+		chip->speed_hz = spi->max_speed_hz;
 		dev_dbg(&spi->dev, "spi->max_speed_hz:%d clk_div:%x cr0:%x",
 			spi->max_speed_hz, clk_div, chip->cr0);
 	}
-- 
2.37.3

